### PR TITLE
Fix completer bug with cycling through options

### PR DESCRIPTION
**Pull request by afshin**
There is a regression in the completer that causes it to behave incorrectly when the user cycles all the way through the options and loops back to the top of the list.

## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/14329

## Code changes

- `getTokenAt` was improved to short-circuit if a token has already been discovered as discussed [in this comment](https://github.com/jupyterlab/jupyterlab/pull/14475#discussion_r1183605532) which was the cause of [issue #14329](https://github.com/jupyterlab/jupyterlab/issues/14329)
- all changes in `completer` package are purely stylistic

## User-facing changes

If the selected item in the completer is an exact match for the text that has already been typed, it will not be erroneously inserted into the text.

**Before**

https://user-images.githubusercontent.com/159529/236578645-248ed66c-8175-47c8-a4b2-c1a2cd2fe44b.mov

**After**

https://user-images.githubusercontent.com/159529/236578667-50655ef7-8572-4615-bded-76bd396586b1.mov

## Backwards-incompatible changes
N/A

### CODE REVIEW

**Comment by trungleduc** [comment]
@afshin i don't have pushing right on this PR

File: packages/completer/src/model.ts [hunk 1]
```diff
@@ -336,6 +336,7 @@ export class CompleterModel implements Completer.IModel {
     }
 
     let { start, end } = cursor;
+    console.log('start', start, 'end', end);
```
Comment (trungleduc) [inline packages/completer/src/model.ts:339]:
```suggestion
```

**Comment by afshin** [comment]
Looks like the test failure is a legitimate issue (either with the test or with the logic in this PR).

File: packages/completer/src/model.ts [hunk 2]
```diff
@@ -368,13 +368,13 @@ export class CompleterModel implements Completer.IModel {
...
     for (let item of items) {
       // See if label matches query string
-      // With ICompletionItems, the label may include parameters, so we exclude them from the matcher.
+      // With ICompletionItems, the label may include parameters,
+      // so we exclude them from the matcher.
       // e.g. Given label `foo(b, a, r)` and query `bar`,
       // don't count parameters, `b`, `a`, and `r` as matches.
       const index = item.label.indexOf('(');
-      let prefix = index > -1 ? item.label.substring(0, index) : item.label;
-      prefix = escapeHTML(prefix);
```
Comment (krassowski) [inline packages/completer/src/model.ts:376]: Yes, the test failures are legit. We absolutely need to keep escaping HTML here.

**Review by krassowski — CHANGES_REQUESTED** [review]
I would appreciate some hints on why you think the changes as proposed here are needed, for example by populating the `Code changes` section :)

> Btw, I can help to fix this regression in a counter PR as it is likely related to refactoring that either I did or I approved - just let me know if help is needed.

Also, I think we should add a test case for the regression.

File: packages/codemirror/src/editor.ts [hunk 3]
```diff
@@ -522,7 +522,7 @@ export class CodeMirrorEditor implements CodeEditor.IEditor {
           if (node.node.firstChild !== null) {
             return true;
           }
-          if (offset >= node.from && offset <= node.to) {
+          if (!token && offset >= node.from && offset <= node.to) {
```
Comment (trungleduc) [inline packages/codemirror/src/editor.ts:525]: Without the check `!token`, after the first matched node, we continue to iterate through the tree and then match other nodes. The returned token is from the last matched node.
In the case of `import`, the first matched token is `{value: 'import', offset: 0, type: 'import'}` but the returned value is `{value: '', offset: 6, type: '⚠'}`, which causes the cycling issue.
I suppose that the first matched node is always the correct one so the `!token` check will ignore all other matched nodes. This fix is tested with other keywords (`class`, `def`,...) and it works as expected.

**Review by afshin — COMMENTED** [review]
bot please update snapshots

**Comment by fcollonval** [comment]
Seems that the bot was not triggered (maybe due to the GitHub troubles).

Bot please update galata snapshots

**Comment by fcollonval** [comment]
Ok this time, the job got triggered: https://github.com/jupyterlab/jupyterlab/actions/runs/4890958620

@krassowski I have a doubt about the failure for search on typing. Do you think the new behavior is legit or not?

**Comment by krassowski** [comment]
I will take a look later, I want to test some edge cases.

**Comment by fcollonval** [comment]
Thanks for the reply Mike. Actually it is not related to this PR, the failure is happening there too e.g.: https://github.com/jupyterlab/jupyterlab/actions/runs/4881704634?pr=14480

**Comment by fcollonval** [comment]
Kicking the CI

File: packages/completer/src/model.ts [hunk 4]
```diff
@@ -368,13 +368,14 @@ export class CompleterModel implements Completer.IModel {
...
-      let match = StringExt.matchSumOfSquares(prefix, query);
+      const text = index > -1 ? item.label.substring(0, index) : item.label;
+      const match =
+        text !== query && StringExt.matchSumOfSquares(escapeHTML(text), query);
```
[the remark this answers is not in the document]
Comment (afshin) [reply packages/completer/src/model.ts]: Thank you!

**Comment by afshin** [comment]
bot please update galata snapshots

**Review by fcollonval — APPROVED** [review]
Thanks @afshin; it works good on Binder

I left this one opened for Mike to review it as the final version could be included any time (no API changes).

**Review by trungleduc — APPROVED** [review]

**Comment by krassowski** [comment]
I updated the `Code changes` section as many code changes in the PR are purely stylistic. I will push a test case for the relevant change.

As for whether we should or should not show the item identical to current token: this is something we could explore indeed; as some auto-insertion logic may depend on it I would not rush to change in before 4.0 though.

File: packages/completer/src/widget.ts [hunk 5]
```diff
@@ -522,34 +522,31 @@ export class Completer extends Widget {
...
-      // Clamp to the length of the list.
-      this._activeIndex = Math.min(
-        Math.max(0, this._activeIndex),
-        items.length - 1
-      );
     }
 
     active = items[this._activeIndex] as HTMLElement;
-    active.classList.add(ACTIVE_CLASS);
+    active?.classList.add(ACTIVE_CLASS);
```
Comment (krassowski) [inline packages/completer/src/widget.ts:549]: Not sure here: is it be more useful for this to lead to an error in console so that users can report it? If `active` ever gets `null`/`undefined` it would indicate a regression in the logic, and would be more difficult to diagnose if this is silenced, right? Not worth halting this PR on this but just food for thought.

**Review by krassowski — APPROVED** [review]
Thank you all! Good to go from my point of view, but I will someone else merge in case if there are comments on my changes.

File: packages/completer/src/widget.ts (hunk 5 above)
Comment (afshin) [reply packages/completer/src/widget.ts:549]: I think this might not actually be necessary and might have been part of the debugging process of this PR. I'm playing around with it and if I can't generate an error, I'll remove the check. It never existed before.

**Comment by afshin** [comment]
bot please update galata snapshots
